src/functions.rs

```diff
@@ -2933,43 +2916,49 @@ pub fn initialize_terminated<'a>(functions: &'a FunctionMap<'a>) -> (Vec<String>
 
 // Helper function to ressolve the "true" name of a function
 // given the FuncCall (the Function we are looking to get the true cil name),
-// FunctionInfo (for the function in which FuncCall is called) and FunctionMap (all FunctionInfos).
+// Context it is called in and FunctionMap (all FunctionInfos).
 // If the function call needs to be cast this will handle that.  This function will not
 // validate the cast that is done else where.
 // If the function call has a "this" class or no class, it will substitue the class of
 // the caller
 // Lastely if the function is type aliased this function will resolve the alias.
 fn resolve_true_cil_name(
     call: &FuncCall,
-    func_info: &FunctionInfo,
+    context: &BlockContext,
+    file: &SimpleFile<String, String>,
     function_map: &FunctionMap,
 ) -> Result<String, CascadeErrors> {
-    // First if check is to see if we need to cast the call.
-    let original_cil_name = if let Some(cast_class) = &call.cast_name {
-        // This should be mostly safe since we have already looked at castables in prevalidate
-        get_cil_name(Some(cast_class), &call.name)
-    }
-    // If we are calling something with this or None it must be in the same class
-    // so hand place the class name
-    else if call.class_name.as_ref() == Some(&CascadeString::from("this"))
-        || call.class_name.is_none()
-    {
-        if let FunctionClass::Type(class) = func_info.class {
-            get_cil_name(Some(&class.name), &call.name)
-        } else {
-            return Err(CascadeErrors::from(
-                ErrorItem::make_compile_or_internal_error(
-                    "Could not determine class for 'this.' function call",
-                    Some(func_info.declaration_file),
-                    call.get_name_range(),
-                    "Perhaps you meant to place the function in a resource or domain?",
-                ),
-            ));
-        }
-    } else {
-        call.get_cil_name()
-    };
-    // Finally deal with alias
+    // The double as_ref() is kind of weird, but I think it's correct.  Option<T>::as_ref() does
+    // &T, rather than T.as_ref().  Since Cascade has implemented the as_ref() trait, we need to
+    // call it explicitly
+    // convert_arg_this() handles this.*, then we handle a bare "this", since we'll be combining it
+    // with a function name ourselves
+    let mut true_call_class = context.convert_arg_this(
+        call.cast_name.as_ref().map(|s| s.as_ref()).unwrap_or(
+            call.class_name
+                .as_ref()
+                .map(|s| s.as_ref())
+                .unwrap_or("this"),
+        ),
+    );
+    if &true_call_class == "this" {
+        true_call_class = match context.get_parent_type_name() {
```

Comment:
Do we need to worry about casting here?  I don't think we do but I want to double check.

Reply:
Didn't we already handle casting above?  Is there some casting scenario in particular that you're concerned about?

Reply:
Sorry miss-read line 2937, thought it said class_name again and thats what your double as_ref comment was about.  You are correct cast should be covered.